Bluetooth: Mesh: Fix unable sent mesh message

LingaoM opened the pull request:
In PR (https://github.com/zephyrproject-rtos/zephyr/pull/58723) has introduce another bug, that, the flag `ADV_FLAG_PROXY` set before actually enabled.

When ctx:: BT RX call `schedule_send` will atomic_test_and_clear `ADV_FLAG_PROXY`, but at this time, the proxy advertising will not at advertising state, maybe in update params or set adverting data phase,

so that, call `bt_le_ext_adv_stop` will nothing, and then call `k_work_reschedule` --> `send_pending_adv`(at this time, the proxy advertising actually enabled, but the upper layer clear proxy flags), cause latest advertising unable start, because unable in advertising state to update params(`-EINVAL`).

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/58721

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -325,8 +325,17 @@ static bool schedule_send(struct bt_mesh_ext_adv *adv)
+	 * otherwise we can only postpone until all hci cmd of the Proxy are completed.
+	 */
+	if (atomic_test_bit(adv->flags, ADV_FLAG_PROXY) &&
+	    atomic_test_bit(adv->instance->flags, BT_ADV_ENABLED)) {
```
Is there a way to not access internal fields of `bt_le_ext_adv`? It would be better if we don't add this dependency to mesh code.

LingaoM replied:
yeah，I'm not found same function todo this.

PavelVPV replied:
An ugly suggestion would be to add `ADV_FLAG_PROXY_SCHEDULED`, set it before `bt_mesh_adv_gatt_send()`, and then check in `connected` callback (I guess this is the only place where it is actually needed).

LingaoM committed:
Bluetooth: Mesh: Fix unable sent mesh message

In PR (#58723) has introduce another bug, that,
the flag ADV_FLAG_PROXY set before actually enabled.

When ctx:: BT RX call schedule_send will atomic_test_and_clear
ADV_FLAG_PROXY, but at this time, the proxy advertising will
not at advertising state, maybe in update params or set adverting
data phase,

so that, call bt_le_ext_adv_stop will nothing, and then call
k_work_reschedule --> send_pending_adv(at this time, the proxy
advertising actually enabled, but the upper layer clear proxy flags),
cause latest advertising unable start, because unable in advertising
state to update params(-EINVAL).

Fixes: #58721

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -327,6 +330,7 @@ static bool schedule_send(struct bt_mesh_ext_adv *adv)
 
 	if (atomic_test_and_clear_bit(adv->flags, ADV_FLAG_PROXY)) {
 		(void)bt_le_ext_adv_stop(adv->instance);
+		atomic_clear_bit(adv->flags, ADV_FLAG_PROXY_START);
```
Would it make sense to clear `ADV_FLAG_PROXY_START` before calling `bt_le_ext_adv_stop`? E.g. if connection happens during scheduling that is caused by `bt_le_ext_adv_stop` (not sure if this possible though)? This would prevent from calling `schedule_send` from `connected` callback.

LingaoM commented:
@Andrewpini @alxelax CC

alxelax reviewed (approved):
It will be nice to have bsim test that covers proxy states. Right now, it is difficult to say we got or we didn't some regression after shuffling states forth and back.
